File: packages/gbnf-experiment/data/20260909T182714Z_2643e4be/ported_implementation/src/GBNF.ts
import { Graph } from "./grammar-graph/graph.js";
import { ParseState } from "./grammar-graph/parse-state.js";
import { buildRuleStack } from "./grammar-parser/build-rule-stack.js";
import { GrammarParseError, RulesBuilder } from "./rules-builder/index.js";
import { ValueError } from "./utils/errors/python-errors.js";

export const GBNF = (grammar: string, initialString = ""): ParseState => {
  if (typeof grammar !== "string") {
    throw new ValueError("grammar must be a string");
  }

  if (typeof initialString !== "string") {
    throw new ValueError("input must be a string");
  }

  const rulesBuilder = new RulesBuilder(grammar);
  const { rules, symbolIds } = rulesBuilder;
  if (rules.length === 0) {
    throw new GrammarParseError(grammar, 0, "No rules were found");
  }
  // as in the reference implementation, a grammar without a root symbol raises
  // a KeyError from this lookup before the check below can report on it.
  if (symbolIds.get("root") === null) {
    throw new GrammarParseError(
      grammar,
      0,
      "Grammar does not contain a 'root' symbol",
    );
  }
  const rootId: number = symbolIds.get("root");

  const stackedRules = rules.map((rule) => buildRuleStack(rule));
  const graph = new Graph(grammar, stackedRules, rootId);
  return new ParseState(graph, graph.add(initialString));
};

File: proposed-projects/round-trip-experiment/reverse/20260909T022216Z_19e0fe30/20260910T055411Z_259d4a57/ported_implementation/gbnf/GBNF.py
from .grammar_graph.graph import Graph
from .grammar_graph.parse_state import ParseState
from .grammar_parser.build_rule_stack import build_rule_stack
from .rules_builder.rules_builder import RulesBuilder
from .utils.errors import GrammarParseError


def GBNF(grammar: str, initial_string: str = "") -> ParseState:
    if not isinstance(grammar, str):
        raise ValueError("grammar must be a string")

    if not isinstance(initial_string, str):
        raise ValueError("input must be a string")

    rules_builder = RulesBuilder(grammar)
    rules = rules_builder.rules
    symbol_ids = rules_builder.symbol_ids
    if len(rules) == 0:
        raise GrammarParseError(grammar, 0, "No rules were found")
    root_id = symbol_ids.get("root")
    if root_id is None:
        raise GrammarParseError(
            grammar,
            0,
            "Grammar does not contain a 'root' symbol",
        )

    stacked_rules = [build_rule_stack(rule) for rule in rules]
    graph = Graph(grammar, stacked_rules, root_id)
    return ParseState(graph, graph.add(initial_string))
